test/modules/GenericMqttTester/src/tester.rs

```diff
@@ -29,42 +70,211 @@ pub struct MessageTester {
+                send_initial_messages(self.publish_handle.clone(), self.message_loop_shutdown_recv)
+                    .instrument(info_span!("initiation message loop")),
+            );
+
+            tasks.push(message_loop);
+        }
+
+        info!("waiting for tasks to exit");
+        let (exited, _, join_handles) = select_all(tasks).await;
+        exited.map_err(MessageTesterError::WaitForShutdown)??;
+        message_channel_shutdown.shutdown().await?;
+        for handle in join_handles {
+            handle
+                .await
+                .map_err(MessageTesterError::WaitForShutdown)??;
+        }
+
+        Ok(())
     }
 
-    pub fn shutdown_handle(&self) -> ShutdownHandle {
+    pub fn shutdown_handle(&self) -> MessageTesterShutdownHandle {
         self.shutdown_handle.clone()
     }
+
+    async fn subscribe(
+        mut client_sub_handle: UpdateSubscriptionHandle,
+        settings: Settings,
+    ) -> Result<(), MessageTesterError> {
+        info!("subscribing to test topics");
+        match settings.test_scenario() {
+            TestScenario::Initiate => client_sub_handle
+                .subscribe(SubscribeTo {
+                    topic_filter: BACKWARDS_TOPIC.to_string(),
+                    qos: QoS::AtLeastOnce,
+                })
+                .await
+                .map_err(MessageTesterError::UpdateSubscription)?,
+            TestScenario::Relay => client_sub_handle
+                .subscribe(SubscribeTo {
+                    topic_filter: FORWARDS_TOPIC.to_string(),
+                    qos: QoS::AtLeastOnce,
+                })
+                .await
+                .map_err(MessageTesterError::UpdateSubscription)?,
+        };
+
+        info!("finished subscribing to test topics");
+        Ok(())
+    }
+}
+
+async fn send_initial_messages(
+    mut publish_handle: PublishHandle,
+    mut shutdown_recv: Receiver<()>,
+) -> Result<(), MessageTesterError> {
+    info!("starting message loop");
+
+    let mut seq_num: u32 = 0;
+    loop {
+        info!("publishing message {} to upstream broker", seq_num);
+        let publication = Publication {
+            topic_name: "forwards/1".to_string(),
+            qos: QoS::ExactlyOnce,
+            retain: true,
+            payload: Bytes::from(seq_num.to_string()),
+        };
+
+        let shutdown_recv_fut = shutdown_recv.next();
+        let publish_fut = publish_handle.publish(publication);
+        pin_mut!(publish_fut);
+
+        match future::select(shutdown_recv_fut, publish_fut).await {
+            Either::Left((shutdown, _)) => {
+                info!("received shutdown signal");
+                shutdown.ok_or(MessageTesterError::ListenForShutdown)?;
+                break;
+            }
+            Either::Right((publish, _)) => {
+                publish.map_err(MessageTesterError::Publish)?;
+            }
+        };
+
+        time::delay_for(Duration::from_secs(1)).await;
+
+        seq_num += 1;
```

Comment:
what are the termination condition of the loop? just shutdown signal? I'm just curios if we should worry about overflowing seq_num.

Reply:
since it runs every second it will be 130+ years